policies/codingstyle.txt

```diff
@@ -289,12 +289,69 @@ options.
 
 In source files, separate functions with one blank line.
 
+                Chapter 6.1: Function Arguments
+
+Most API functions in OpenSSL revolve around a specific type (e.g. SSL_CTX,
+EVP_PKEY, X509 etc). When adding new API functions relevant to a given type, the
+following argument ordering should be used:
+
+    rettype TYPE_get_something(TYPE *,[args])
+    rettype TYPE_do_something(TYPE *,[args])
+    TYPE *TYPE_new([args])
+
+Where present the TYPE argument should always be first. In constructors or
+object factory type functions (such as a "fetch" function), the key argument
+that identifies the thing be constructed should be placed first (if such an
+argument exists). For example in a "fetch" function the name of the algorithm
+being fetched would be first.
+
+The remaining argument list should be put into order of importance. Deciding the
+relative importance of arguments is necessarily somewhat subjective, but the
+following guidelines should be considered:
+
+1) Any "context" arguments should be first. If multiple contexts are necessary
+   then they should be in order of most general to most specific. In some cases
+   (typically with an OPENSSL_CTX), a context may be NULL to indicate a default
+   context should be used. This does not lessen its importance. Since such a
+   context is always used, even though a default value may be applied, it does
+   not make it an "optional" argument.
```

Comment:
I believe this statement is fine as long as it concerns the core developers, but is totally false from the PoV of the users of these functions for which we should design our API.

As a C developer, if there is a pointer argument where I can legally pass NULL instead of something I fetched or malloced, that is something I can pass optionally.
From the API user point of view it does not really matter an internal distinction on what the function will do if I pass "NULL-nothing" as the value of that argument: it might pick a global value, it might create a temporary replacement or it might just skip stuff that requires it. Whatever happens inside stays inside, I am passing "nothing" so it is optional to create/fetch a non-NULL reference to pass in that argument. 

If we want to make a distinction between a function where passing NULL means you are passing nothing and one where passing NULL means you are explicitly referring to the default global value, wouldn't we be better served by not using NULL but another define (maybe mapped to NULL, maybe not, but still our documentation and code would pass OSSL_DFLT instead of NULL, thus removing the impression of passing nothing for our users)?

Reply:
I actually think the argument about where "NULL" means its optional or not, is not a particularly helpful one. I could actually drop the text from "This does not lessen its importance..." to the end of that bullet. Its sufficient just to say contexts come first, regardless of whether they are NULL or not.

Reply:
@romen:
> As a C developer, if there is a pointer argument where I can legally pass NULL instead of something I fetched or malloced, that is something I can pass optionally.

That depends on if NULL is a significant value (which it is for a library context) or not.

Reply:
I have removed the section of text I mentioned above.

Reply:
@romen I think the better way to think of this is that conceptually what the argument is referencing is **always** required and that there are two ways it can be provided, one through reference to a value and one through an internal default if the parameter is passed as a special sentinel (NULL). Conceptually it is always present. You could even directly pass in the result of a call to fetch the default. 

Having an API that makes it clear there is always a context  helps the user to understand that it is an important concept to keep in mind - there is always a context conceptually required even if you pass NULL.